### ISSUE TITLE
Url can't be changed for easy reconnect

### ISSUE THREAD

**Issue by benaadams** [dotnet/aspnetcore#5342]
Using a client-side cursor (e.g. `lastId` processed); for a reconnect this wants to be passed back to the server in the query string (as cors doesn't allow custom headers and cors is always applicable to websockets).

However while `signalR.HubConnectionBuilder()` provides easy construction of the connection; the url cannot be changed after building.

This means if the connection disconnects and you need to pass a cursor (e.g. id); the connection has to be rebuilt and all the events re-registered to the new connection.

It would be easier if the url could be updated and only `.start` needed to be called again.

**Comment by muratg** [comment]
cc @halter73 because I saw the word "reconnect" :smiley:

**Comment by halter73** [comment]
This wouldn't affect automatic reconnects. If we want this, we should probably set up a design meeting to determine things like how the URL should be updated (e.g. a property or `.start` param), when it can be updated, etc...

**Comment by anurse** [comment]
Triage decision: Design this in the preview 6 time period.

**Comment by anurse** [comment]
Proposal: Make `Url` settable, but it doesn't take affect until the next `StartAsync`/`start` call. We'd capture the current value in the connection state. We'd do this in the HubConnection layer (since `HttpConnection` is not reusable anyway).

**Comment by benaadams** [comment]
Your proposal is acceptable... {insert "men in black" gif}

**Comment by anurse** [comment]
Still planning to get this done... eventually 😫 

With the reconnect stuff, this should be doable in the `Reconnecting` event as well. @halter73 wants to try throwing `InvalidOperationException` if you try to set it while the connection is active (not reconnecting or disconnected). So sure, give that a whirl :)

**Comment by anurse** [comment]
This is done in Java and JS. The .NET work is blocked on #10872 because we'll be refactoring that logic.

**Comment by rjvdboon** [comment]
> 
> 
> The .NET work is blocked on #10872 because we'll be refactoring that logic.

It seems #10872 is done now

**Comment by mikaelm12** [comment]
Yeah, #10872 came in right at the end of preview 7 so this will now go into the preview 8 milestone. I can go ahead and finish this up.

**Comment by anurse** [comment]
We missed the 3.0 train. We'll revisit for 3.1

**Comment by coonmoo** [comment]
Any news on getting the .NET SignalR client to support changing the url in the reconnecting event?
I see in the JS client this is already supported
